File: C#/P2.2/Generator/CFFKernel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Security.Cryptography;
using System.Text;
using System.Threading.Tasks;
using P2._2;

namespace P2._2.Generator
{
    public class CFFKernel
    {
        public static List<CFFKernel> InitKernals(GenProperties gp) {
            List<CFFKernel> kernels = new List<CFFKernel>();
            for(int r = gp.f_radiusvar.low; r <= gp.f_radiusvar.high; r++) {
                kernels.Add(new CFFKernel(r, gp.f_height));
            }
            return kernels;
        }

        public int r { get; }
        int size, anchor;
        float[,] data;

        public CFFKernel(int radius, float height) {
            if (height < 0) throw new ArgumentException($"CFFKernel received a height of {height}, must be above 0.0!");

            // Process radius
            r = radius;
            size = r * 2 - 1;
            data = new float[size, size];
            anchor = (size - 1) / 2;

            // Compute kernel data (CFF)
            for(int y = 0; y < size; y++) {
                for(int x = 0; x < size; x++) {
                    int s = r * r - (x - anchor) * (x - anchor) - (y - anchor) * (y - anchor);
                    if (s <= 0) continue;
                    data[x, y] = (float)(Math.Sqrt(s) / r) * height;
                }
            }

        }

        public void Apply(Heightmap heightmap, int x, int y) {
            // Translated CFF Corner coordinates
            (int x, int y) p1 = (x - anchor, y - anchor);
            (int x, int y) p2 = (x + anchor, y + anchor);

            if (p1.x >= heightmap.width || p1.y >= heightmap.height || p2.x < 0 || p2.y < 0) return;

            int inset_l = p1.x >= 0 ? 0 : Math.Abs(p1.x);
            int inset_t = p1.y >= 0 ? 0 : Math.Abs(p1.y);
            int inset_r = p2.x < heightmap.width ? 0 : Math.Abs(p2.x - (heightmap.width - 1));
            int inset_b = p2.y < heightmap.height ? 0 : Math.Abs(p2.y - (heightmap.height - 1));

            if (inset_r + inset_l >= size || inset_t + inset_b > size) throw new Exception("Kernel insets exceeded kernel size! Maybe kernel outside map?");

            for(int dy = inset_t; dy < size - inset_b; dy++) {
                for(int dx = inset_l; dx < size - inset_r; dx++) {
                    int hm_x = x - anchor + dx;
                    int hm_y = y - anchor + dy;
                    float v = heightmap[hm_x, hm_y];
                    float hm_v = v + data[dx, dy];
                    heightmap[hm_x, hm_y] = hm_v;
                }
            }
        }
    }
}

File: Python/CFFkernel.py
import numpy as np
import math

def InitKernels(radiusvar, height):
	kernels = []
	for r in range(radiusvar[0], radiusvar[1] + 1):
		kernels.append(CFFKernel(r, height))
	return kernels

class CFFKernel:
	def __init__(self, r, height):
		if height < 0:
			raise Exception("CFFKernel was supplied a height of " + str(height) + ". Must be above 0.0!")

		# process radius
		self._r = math.ceil(r)
		self._size = self._r * 2 - 1 # should be + 1, but the eddge values will always be 0
		self._data = np.zeros((self._size, self._size))
		self._anchor = math.floor((self._size - 1) / 2)

		# compute kernel data (CFF)
		for y in range(0, self._size):
			for x in range(0, self._size):
				s = self._r**2 - (x - self._anchor)**2 - (y - self._anchor)**2
				if s <= 0:
					continue
				self._data[x][y] = (math.sqrt(s) / r) * height

	def Apply(self, heightmap, x, y):
		# corner coordinates
		p1 = (x - self._anchor, y - self._anchor) # top-left corner
		p2 = (x + self._anchor, y + self._anchor) # bottom-right corner
		
		if p1[0] >= heightmap._width or p1[1] >= heightmap._height or p2[0] < 0 or p2[1] < 0:
			print("I: Kernel outside heightmap bounds.")
			return
		
		inset_l = 0 if p1[0] >= 0 else abs(p1[0])
		inset_t = 0 if p1[1] >= 0 else abs(p1[1])
		inset_r = 0 if p2[0] < heightmap._width else (p2[0] - (heightmap._width - 1))
		inset_b = 0 if p2[1] < heightmap._height else (p2[1] - (heightmap._height - 1))

		if (inset_r + inset_l) >= self._size or (inset_t + inset_b) >= self._size:
			raise Exception("Kernel insets exceeded kernel size! Maybe kernel outside map?")
	
		for dy in range(inset_t, self._size - inset_b):
			for dx in range(inset_l, self._size - inset_r):
				hm_x = x - self._anchor + dx
				hm_y = y - self._anchor + dy
				v = heightmap.GetValue(hm_x, hm_y)
				#hm_v = v + (1 - v) * self._data[dx][dy]
				hm_v = v + self._data[dx][dy]
				heightmap.SetValue(hm_x, hm_y, hm_v)